File: packages/platform-python/src/platform_python/errormodel.py
"""RFC 7807 error envelope shared by every Python service.

Mirrors `platform-spring-boot-error` (Kotlin) and `platform-go/errormodel`
(Go). The body shape is normative:

.. code-block:: json

    {
      "type":       "https://platform.trips-enjoy.com/errors/<code-kebab>",
      "title":      "<Human title>",
      "status":     503,
      "detail":     "<service-specific detail>",
      "instance":   "<request path>",
      "code":       "DEPENDENCY_UNAVAILABLE",
      "traceId":    "<otel trace_id>",
      "spanId":     "<otel span_id>",
      "timestamp":  "RFC3339",
      "errors":     [{"field": "amount", "message": "must be > 0", "code": "MIN_VALUE"}],
      "downstream": {"service": "payment-service", "code": "CIRCUIT_OPEN", "status": 503,
                    "traceId": "...", "latency_ms": 17, "attempt": 1}
    }
"""
from __future__ import annotations

from datetime import datetime, timezone
from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ErrorCode(str, Enum):
    """SCREAMING_SNAKE_CASE machine identifier per the platform catalog."""

    VALIDATION_FAILED = "VALIDATION_FAILED"
    UNAUTHENTICATED = "UNAUTHENTICATED"
    FORBIDDEN = "FORBIDDEN"
    NOT_FOUND = "NOT_FOUND"
    CONFLICT = "CONFLICT"
    IDEMPOTENCY_KEY_REUSED = "IDEMPOTENCY_KEY_REUSED"
    RATE_LIMITED = "RATE_LIMITED"
    BUSINESS_RULE_VIOLATION = "BUSINESS_RULE_VIOLATION"
    STATE_INVALID = "STATE_INVALID"
    INTERNAL_ERROR = "INTERNAL_ERROR"
    DEPENDENCY_UNAVAILABLE = "DEPENDENCY_UNAVAILABLE"
    DEPENDENCY_TIMEOUT = "DEPENDENCY_TIMEOUT"
    BAD_GATEWAY = "BAD_GATEWAY"
    CIRCUIT_OPEN = "CIRCUIT_OPEN"
    BULKHEAD_FULL = "BULKHEAD_FULL"


# Canonical HTTP status per ErrorCode. Mirrors the Kotlin and Go maps.
_HTTP_STATUS: dict[ErrorCode, int] = {
    ErrorCode.VALIDATION_FAILED: 400,
    ErrorCode.UNAUTHENTICATED: 401,
    ErrorCode.FORBIDDEN: 403,
    ErrorCode.NOT_FOUND: 404,
    ErrorCode.CONFLICT: 409,
    ErrorCode.IDEMPOTENCY_KEY_REUSED: 422,
    ErrorCode.BUSINESS_RULE_VIOLATION: 422,
    ErrorCode.RATE_LIMITED: 429,
    ErrorCode.BAD_GATEWAY: 502,
    ErrorCode.CIRCUIT_OPEN: 503,
    ErrorCode.DEPENDENCY_UNAVAILABLE: 503,
    ErrorCode.BULKHEAD_FULL: 503,
    ErrorCode.DEPENDENCY_TIMEOUT: 504,
    ErrorCode.INTERNAL_ERROR: 500,
    ErrorCode.STATE_INVALID: 409,
}


def http_status_for(code: ErrorCode) -> int:
    return _HTTP_STATUS.get(code, 500)


class FieldError(BaseModel):
    """One row in the errors[] array (validation failures)."""

    field: str
    message: str
    code: str | None = None


class Downstream(BaseModel):
    """Identifies the originating service for an error that crossed a boundary."""

    service: str
    code: str | None = None
    status: int
    traceId: str | None = None
    latency_ms: int | None = None
    attempt: int | None = None


class ErrorEnvelope(BaseModel):
    """RFC 7807 + platform-extension response body."""

    model_config = ConfigDict(populate_by_name=True)

    type: str
    title: str
    status: int
    detail: str
    instance: str
    code: ErrorCode
    traceId: str | None = None
    spanId: str | None = None
    timestamp: datetime
    errors: list[FieldError] | None = Field(default=None)
    downstream: Downstream | None = Field(default=None)

    @classmethod
    def build(
        cls,
        code: ErrorCode,
        detail: str,
        instance: str,
        trace_id: str | None = None,
        span_id: str | None = None,
        errors: list[FieldError] | None = None,
        downstream: Downstream | None = None,
    ) -> "ErrorEnvelope":
        return cls(
            type=f"https://platform.trips-enjoy.com/errors/{code.value.lower().replace('_', '-')}",
            title=_titleize(code),
            status=http_status_for(code),
            detail=detail,
            instance=instance,
            code=code,
            traceId=trace_id,
            spanId=span_id,
            timestamp=datetime.now(timezone.utc),
            errors=errors,
            downstream=downstream,
        )


def _titleize(code: ErrorCode) -> str:
    return " ".join(p.title() for p in code.value.split("_"))


def to_fastapi_response(envelope: ErrorEnvelope) -> dict[str, Any]:
    """Render the envelope as a JSON-serialisable dict for FastAPI."""
    return envelope.model_dump(mode="json", exclude_none=True)

File: packages/platform-go/errormodel/errormodel.go
// Package errormodel defines the platform-wide RFC 7807 error envelope
// shared by every Go service (api-gateway, chat-service, file-service,
// geolocation-service). The JSON shapes are normative — matching the
// Spring Boot `platform-spring-boot-error` module and the FastAPI
// `platform-python` package one-for-one.
//
// Source of truth: docs/shared/CONVENTIONS.md §1.
//
// The envelope is:
//
//	{
//	  "type":       "https://platform.trips-enjoy.com/errors/<code-kebab>",
//	  "title":      "<Human title>",
//	  "status":     503,
//	  "detail":     "<service-specific detail>",
//	  "instance":   "<request path>",
//	  "code":       "DEPENDENCY_UNAVAILABLE",
//	  "traceId":    "<otel trace_id>",
//	  "spanId":     "<otel span_id>",
//	  "timestamp":  "RFC3339",
//	  "errors":     [{"field": "...", "message": "...", "code": "MIN_VALUE"}],
//	  "downstream": {"service": "...", "code": "...", "status": 503,
//	                "traceId": "...", "latency_ms": 17, "attempt": 1}
//	}
package errormodel

import (
	"strconv"
	"strings"
	"time"
)

// Code is the SCREAMING_SNAKE_CASE machine identifier per the platform
// catalog. The list mirrors `ErrorCode` in
// `platform-spring-boot-error/src/main/kotlin/.../ErrorCode.kt`.
type Code string

const (
	CodeValidationFailed      Code = "VALIDATION_FAILED"
	CodeUnauthenticated       Code = "UNAUTHENTICATED"
	CodeForbidden             Code = "FORBIDDEN"
	CodeNotFound              Code = "NOT_FOUND"
	CodeConflict              Code = "CONFLICT"
	CodeIdempotencyKeyReused  Code = "IDEMPOTENCY_KEY_REUSED"
	CodeRateLimited           Code = "RATE_LIMITED"
	CodeBusinessRuleViolation Code = "BUSINESS_RULE_VIOLATION"
	CodeStateInvalid          Code = "STATE_INVALID"
	CodeInternalError         Code = "INTERNAL_ERROR"
	CodeDependencyUnavailable Code = "DEPENDENCY_UNAVAILABLE"
	CodeDependencyTimeout     Code = "DEPENDENCY_TIMEOUT"
	CodeBadGateway            Code = "BAD_GATEWAY"
	CodeCircuitOpen           Code = "CIRCUIT_OPEN"
	CodeBulkheadFull          Code = "BULKHEAD_FULL"
)

// HTTPStatus is the canonical HTTP status for each platform code.
func (c Code) HTTPStatus() int {
	switch c {
	case CodeValidationFailed:
		return 400
	case CodeUnauthenticated:
		return 401
	case CodeForbidden:
		return 403
	case CodeNotFound:
		return 404
	case CodeConflict:
		return 409
	case CodeIdempotencyKeyReused, CodeBusinessRuleViolation:
		return 422
	case CodeRateLimited:
		return 429
	case CodeBadGateway:
		return 502
	case CodeCircuitOpen, CodeDependencyUnavailable, CodeBulkheadFull:
		return 503
	case CodeDependencyTimeout:
		return 504
	case CodeStateInvalid:
		return 409
	default:
		return 500
	}
}

// FieldError is one row in the `errors[]` array (validation failures).
type FieldError struct {
	Field   string `json:"field"`
	Message string `json:"message"`
	Code    string `json:"code,omitempty"`
}

// Downstream identifies the originating service for an error that
// crossed a service boundary.
type Downstream struct {
	Service   string `json:"service"`
	Code      string `json:"code,omitempty"`
	Status    int    `json:"status"`
	TraceID   string `json:"traceId,omitempty"`
	LatencyMs int64  `json:"latency_ms,omitempty"`
	Attempt   int    `json:"attempt,omitempty"`
}

// Envelope is the RFC 7807 + platform-extension response body.
type Envelope struct {
	Type      string       `json:"type"`
	Title     string       `json:"title"`
	Status    int          `json:"status"`
	Detail    string       `json:"detail"`
	Instance  string       `json:"instance"`
	Code      Code         `json:"code"`
	TraceID   string       `json:"traceId,omitempty"`
	SpanID    string       `json:"spanId,omitempty"`
	Timestamp string       `json:"timestamp"`
	Errors    []FieldError `json:"errors,omitempty"`
	Downstream *Downstream `json:"downstream,omitempty"`
}

// now is override-able in tests.
var now = func() time.Time { return time.Now().UTC() }

// New constructs an Envelope for a business error.
//
//   - code:        the platform ErrorCode
//   - detail:      service-specific human message
//   - instance:    the request path (e.g. r.URL.Path)
//   - traceID:     OTel trace_id (may be empty)
//   - spanID:      OTel span_id (may be empty)
func New(code Code, detail, instance, traceID, spanID string) Envelope {
	return Envelope{
		Type:      "https://platform.trips-enjoy.com/errors/" + kebab(code),
		Title:     titleize(code),
		Status:    code.HTTPStatus(),
		Detail:    detail,
		Instance:  instance,
		Code:      code,
		TraceID:   traceID,
		SpanID:    spanID,
		Timestamp: now().Format(time.RFC3339Nano),
	}
}

// NewWithDownstream constructs an Envelope attached to a downstream
// originator block.
func NewWithDownstream(code Code, detail, instance, traceID, spanID string, ds *Downstream) Envelope {
	e := New(code, detail, instance, traceID, spanID)
	e.Downstream = ds
	return e
}

// NewValidation constructs an Envelope with one or more field errors
// (status 400).
func NewValidation(detail, instance, traceID, spanID string, fieldErrors []FieldError) Envelope {
	e := New(CodeValidationFailed, detail, instance, traceID, spanID)
	e.Errors = fieldErrors
	return e
}

func kebab(c Code) string {
	return strings.ReplaceAll(strings.ToLower(string(c)), "_", "-")
}

func titleize(c Code) string {
	parts := strings.Split(strings.ToLower(string(c)), "_")
	for i, p := range parts {
		parts[i] = strings.ToUpper(p[:1]) + p[1:]
	}
	return strings.Join(parts, " ")
}

// IntFromContext reads an integer from a context field (helpers for
// downstream error payloads).
func IntFromContext(s string, fallback int) int {
	if s == "" {
		return fallback
	}
	n, err := strconv.Atoi(s)
	if err != nil {
		return fallback
	}
	return n
}
